plugins/state_monitor.py

```diff
@@ -92,77 +100,217 @@ def execute(self, input_data=None):
-            if (state_now >= 12) and (state_now != state_pre):
-                msg = f"Critical failure with charge controller. LED state: {state_now}."
-                self.logger.info(msg)
-                self.send_message(msg)
-
-        # If expression evaluation fails, presumably due to bad data values
-        except Exception as e:
-            self.logger.error(
-                "%s evaluating in %s on step %s: %s",
-                type(e).__name__, type(self), self.name, e)
-            self.logger.info("Error details:", exc_info=True)
-            for pretty_name, data in [("Current", input_data),
-                                      ("Previous", self._previous_data)]:
-                self.logger.info(
-                    "%s data: %r", pretty_name,
-                    {key: str(value) for key, value in data.items()})
+        # TODO detect this more reliably by checking array_fault and load_fault bitfields are non-zero,
 
         # Update state for next pass through the pipeline
         self._previous_data = input_data
 
         # Pass through the input for consumption by any further steps
         return input_data
 
+    def now_then(self, input_data, key):
+        """
+        Simple method to extract the current (now) value and previous (then)
+        value for the data structure passed around.
+
+        Useful for getting any value from `input_data`, since this we turn
+        string 'NA's to numeric NaNs.
+
+        Parameters
+        ----------
+        input_data : Mapping[str, DataValue]
+            Same as argument of `execute`.
+        key : str
+            The key of `input_data` you want the value of.
+
+        Returns
+        -------
+        now_then_values : tuple
+            Two element tuple of the (now value, then value)
+
+        """
+
+        now_val = input_data[key].value
+        then_val = self._previous_data[key].value
+
+        if now_val == 'NA':
+            now_val = nan
+
+        if then_val == 'NA':
+            then_val = nan
+
+        return now_val, then_val
+
+    def log_error(self, input_data, exception_inst):
+        """
+        Log an error.
+
+        Use this when catching Exceptions, especially in the methods of the class.
+
+        Parameters
+        ----------
+        input_data : Mapping[str, DataValue]
+            Same as argument of `execute`.
+
+        exception_inst : Exception
+            The exception you wish to log.
+
+        """
+
+        self.logger.error(
+            "%s evaluating in %s on step %s: %s",
+            type(exception_inst).__name__, type(self), self.name, exception_inst)
+        self.logger.info("Error details:", exc_info=True)
+        for pretty_name, data in [("Current", input_data),
+                                  ("Previous", self._previous_data)]:
+            self.logger.info(
+                "%s data: %r", pretty_name,
+                {key: str(value) for key, value in data.items()})
```

Comment:
```suggestion
                "%s data: %r", pretty_name, data)
```

I'm not totally sure why we're casting every value manually to `str()` here—this hinders diagnosing a `TypeError` with the current state monitor code attempting to compare a string input data value with an `int` constant value in the low space check (as its less clear which or why certain inputs are strings rather than numeric values):

<details>

<summary>Full error for reference</summary>

```
2023-05-01 21:34:35.930 | ERROR | Telemetry Process | plugins.state_monitor.StateMonitor | TypeError evaluating in on step State Monitor: '<' not supported between instances of 'str' and 'int' (T+107521260 ms)
2023-05-01 21:34:35.930 | INFO | Telemetry Process | plugins.state_monitor.StateMonitor | Error details:
Traceback (most recent call last):
File "/home/pi/dev/mjolnir-hamma/plugins/state_monitor.py", line 124, in execute
if (space_now < self.low_space) and (space_pre > self.low_space):
TypeError: '<' not supported between instances of 'str' and 'int' (T+107521260 ms)
2023-05-01 21:34:35.931 | INFO | Telemetry Process | plugins.state_monitor.StateMonitor | Current data: {'time': '2023-05-01 21:34:34.702969+00:00', 'runtime': '107520.001', 'adc_vb_f': '13.034', 'adc_va_f': '29.044', 'adc_vl_f': '13.077', 'adc_ic_f': '0.5', 'adc_il_f': '0.22', 't_hs': '33', 't_batt': '33', 't_amb': '33', 't_rts': '25', 'charge_state': '5', 'array_fault': '0', 'vb_f': '13.037', 'vb_ref': '13.588', 'ahc_r': '29.5', 'ahc_t': '29.5', 'kwhc': '0.3', 'load_state': '1', 'load_fault': '0', 'v_lvd': '10.999', 'ahl_r': '4.6', 'ahl_t': '4.6', 'hourmeter': '3512', 'alarm': '1', 'dip_switch': '11', 'led_state': '6', 'power_out': '6.507', 'sweep_vmp': '29.092', 'sweep_pmax': '7.821', 'sweep_voc': '38.239', 'vb_min_daily': '12.732', 'vb_max_daily': '13.962', 'ahc_daily': '29.5', 'ahl_daily': '4.6', 'array_fault_daily': '0', 'load_fault_daily': '0', 'alarm_daily': '262145', 'vb_min': '6.47', 'vb_max': '15.076', 'ping': '1', 'sequence_count': 'NA', 'timestamp': 'NA', 'crc_errors': 'NA', 'valid_packets': 'NA', 'bytes_read': 'NA', 'bytes_written': 'NA', 'bytes_remaining': 'NA', 'packets_sent': 'NA', 'packets_dropped': 'NA'} (T+107521261 ms)
2023-05-01 21:34:35.932 | INFO | Telemetry Process | plugins.state_monitor.StateMonitor | Previous data: {'time': '2023-05-01 21:33:34.702974+00:00', 'runtime': '107460.001', 'adc_vb_f': '13.083', 'adc_va_f': '29.202', 'adc_vl_f': '13.071', 'adc_ic_f': '0.519', 'adc_il_f': '0.312', 't_hs': '33', 't_batt': '33', 't_amb': '33', 't_rts': '25', 'charge_state': '5', 'array_fault': '0', 'vb_f': '13.04', 'vb_ref': '13.588', 'ahc_r': '29.5', 'ahc_t': '29.5', 'kwhc': '0.3', 'load_state': '1', 'load_fault': '0', 'v_lvd': '10.999', 'ahl_r': '4.6', 'ahl_t': '4.6', 'hourmeter': '3512', 'alarm': '1', 'dip_switch': '11', 'led_state': '6', 'power_out': '6.764', 'sweep_vmp': '29.242', 'sweep_pmax': '8.38', 'sweep_voc': '38.403', 'vb_min_daily': '12.732', 'vb_max_daily': '13.962', 'ahc_daily': '29.5', 'ahl_daily': '4.6', 'array_fault_daily': '0', 'load_fault_daily': '0', 'alarm_daily': '262145', 'vb_min': '6.47', 'vb_max': '15.076', 'ping': '1', 'sequence_count': 'NA', 'timestamp': 'NA', 'crc_errors': 'NA', 'valid_packets': 'NA', 'bytes_read': 'NA', 'bytes_written': 'NA', 'bytes_remaining': 'NA', 'packets_sent': 'NA', 'packets_dropped': 'NA'} (T+107521262 ms) 
```

</details>

Reply:
I don't have any issue with this change. The only time I remember it being "a thing" was when we have binary data in the dict. I'm going to accept so me and @jcburchfield can get this implemented.